### ISSUE TITLE
TypeVar with a default value followed by a TypeVar without one, reports an unreasonable error.

### ISSUE THREAD

**Issue by GreyElaina** [microsoft/pyright#5444]
**Describe the bug**
```py
from __future__ import annotations

from typing_extensions import TypeVar

T = TypeVar("T")
D = TypeVar("D", default=str)

def x(a: D, b: T) -> T: 
	#^: report ""T" cannot appear after "D" in the list of type parameters because it does not have a default type."
    return b
```

**To Reproduce**
upper code.

**Expected behavior**
pyright doesn't complain anything.

**Code or Screenshots**

**VS Code extension or command-line**
latest pylance, with the good chinese translation.

**Additional context**

**Comment by debonte** [comment]
[PEP 696](https://peps.python.org/pep-0696/#default-ordering-and-subscription-rules) specifies that this is an error:

> The order for defaults should follow the standard function parameter rules, so a `TypeVarLike` with no `default` cannot follow one with a `default` value. Doing so should ideally raise a `TypeError` in `typing._GenericAlias`/`types.GenericAlias`, and a type checker should flag this an error.

So Pyright is behaving correctly here.

**Comment by GreyElaina** [comment]
> [PEP 696](https://peps.python.org/pep-0696/#default-ordering-and-subscription-rules) specifies that this is an error:
> 
> > The order for defaults should follow the standard function parameter rules, so a `TypeVarLike` with no `default` cannot follow one with a `default` value. Doing so should ideally raise a `TypeError` in `typing._GenericAlias`/`types.GenericAlias`, and a type checker should flag this an error.
> 
> So Pyright is behaving correctly here.

uh, could i have a problem? : I believe this paragraph refers to cases like class Test(Generic[A, B, C, ...]), rather than type inference for functions or callables.
or i read this pep first?

**Comment by debonte** [comment]
Oh, sorry, yes I see what you're saying. I wonder if

```python
def x(a: D, b: T) -> T:
```

is bring interpreted as:

```python
def x[D, T](a: D, b: T) -> T:
```

Now that [PEP 695](https://peps.python.org/pep-0695/#summary-examples) had been approved, I wonder if you can fix this by explicitly specifying the order of the TypeVars as:

```python
def x[T, D](a: D, b: T) -> T:
```

This is just a hunch though. I'm on my phone and can't try it at the moment.

**Comment by GreyElaina** [comment]
yes, i asked bing chat, and i found we can "sort" these variables, however i donot know if it's a undecidable problem

**Comment by debonte** [comment]
> however i donot know if it's a undecidable problem

@erictraut, when the TypeVar order for a function is not explicitly specified, could Pyright sort the TypeVars in the implicit list so those with defaults are at the front?

**Comment by erictraut** [comment]
Pyright is doing the correct thing here. The order of type parameterization for a function is determined by the order in which the type variables appear (left to right). Prior to PEP 695, there was no way to specify this order explicitly.

Sorting the type parameters is not part of the spec, so it would be improper for pyright to do it. It could lead to inconsistencies between type checkers. It would also preclude the possibility of explicit function specialization in the future (e.g. `foo[int, str](1, "")`). This was part of earlier drafts of PEP 695, but I removed it as part of a simplification pass. There are active discussions about creating a new PEP to add back this feature. So I think it's unlikely that the typing community would endorse an "implicit sorting".

I'll note that PEP 696 has not yet been accepted, so its support within pyright is provisional. Please don't use it in any production code until it has been officially accepted (which will presumably happen as part of the Python 3.13 cycle).
